### ISSUE TITLE
Feature request: Optional range parameter for download request

### ISSUE THREAD

**Issue by sondree** [Backblaze/B2_Command_Line_Tool#250]
Hi! 

Developer of B2Fuse here. An idea was put forth to use B2 CLI Python API as backend for B2Fuse to handle integration against B2. This is an excellent idea but requires an additional feature from B2 CLI. In order to use B2 CLI Python API in B2Fuse the download call needs to be able to request part of a file. This should be implemented as an optional feature, as in some cases it will also be necessary to request entire files.

Would this be possible to add? 

Best 
Sondre

**Comment by ppolewicz** [comment]
I would see that as:
- definition of new `range` parameter, which would be a two-element ordered collection (usually a tuple) consisting of two integers, as in [b2 API documentation](https://www.backblaze.com/b2/docs/b2_download_file_by_id.html)
- addition of optional `range` parameter for `Bucket.download_file_by_name()`
- addition of optional `range` parameter for `Bucket.download_file_by_id()`
- addition of optional `range` parameter for `B2Api.download_file_by_id()`
- addition of optional `range` parameter for `raw_api.download_*` methods
- download destination will be largely unchanged (except for hash checking, if it is in the way)

@svonohr @bwbeach will that be ok?

**Comment by bwbeach** [comment]
Sounds good to me.

**Comment by ppolewicz** [comment]
@sondree after further analysis it was discovered that reading an encrypted / compressed file might be slightly more complicated. An *open, (seek&read many times), close* workflow might make more sense. If *b2_fuse* shall later support such files, then we shall provide a consistent interface already.

Would it be ok for you to call
```python
f = api.read_file_by_id(fileId)
f.seek(position)
f.read(requested_read_size+read_ahead_buffer)
```
with
```python
f.close()
```
called later, when the user stops reading the file and you recycle resources?

**Comment by sondree** [comment]
Sorry for the late reply.

Yes, it would not be a bad thing if your "file" objects acted more like regular pythonic files. In fact as I am developing towards first release using your backend I see that there is some room for improvements. For instance, is there a reason why deletion of files is delegated to the session object (and not the bucket)?

**Comment by ppolewicz** [comment]
You should never use the session directly. (in fact it should have been hidden from the API, we need to do it before v1.0...).

There is no good reason it is not available in the Bucket class (it has to be available in B2Api because you can delete by id, not knowing in which bucket the file is). Could you please create a new issue and specify how you'd like the API to look like (delete by file id or by file name? the second one is more expensive)

There is advantage in having your implementation use the "file" object from the beginning. I'll work on that.

**Comment by sondree** [comment]
How would a "file" object handle large files? In enabling encryption would you need access to the entire file? The use case for B2Fuse of reading parts of files is to avoid having to download the entire file before starting to return data to user applications. If you then have to download the entire file to make encryption work, that would bring us back to square one. 

File like objects present a clear and concise way of interacting with data you provide. While at the same time I am unsure if it should be Python B2 API's job to handle the allocation of memory or disk space to store these files. Would it not make more sense that Python B2 API tried to be as minimal, fast and transparent as possible? Maybe this is a non-issue if you don't need to download the entire file to decrypt. In any case it would be most important that the overhead remain low, I would look into compression and encryption working on blocks to alleviate these issues if you have not already.

**Comment by ppolewicz** [comment]
We plan to minimize the amount of data that needs to be downloaded by making it work on blocks. Going for the same overhead target as you do :)

In case of encryption, we need some kind of header, so we'll get it once and keep it in memory until you close the file handle. In case of compression, it is tricky, but if a file-like object is used, we can do much better than with a crude approach.

Large files should be handled completely transparently.

What I fear is that everyone that tries to implement fuse, web api, QT interface or whatever, will need to implement all the logic that you described. Therefore if we'd keep B2 API layer very thin, someone would have to create another library that would wrap it into something usable. That does not make much sense. It is better if we do it once in the Python B2 API layer and everyone just uses it in a civilized way. One of good examples of such improvement is retrying uploads when something goes wrong.

The allocation of memory is a concern. Can you imagine an interface which would download data from B2Api and hand it over to your layer? We have `b2.download_dest.DownloadDestBytes`. Is it good for you or do you need something else?

**Comment by sondree** [comment]
Yes, something like DownloadDestBytes might do. I would have to look into how to use that. Essentially B2Fuse should be responsible for managing all memory and disk consumption excluding a small working buffer for the features mentioned. 

I have already had several people wanting to use B2Fuse but being put off by memory constraints, as such I am looking into the use of disk space as an alternative. Essentially there needs to be a way for me to take control of the memory used, maybe it doesn't have to be all that complicated. As long as the file object does not have a "hidden" copy of large parts of the file stored, I think it should be fine.

**Comment by ppolewicz** [comment]
My idea was to make it so our layer performs reading and hands off data to your layer, which can then decide what do do with it. This was the idea behind `AbstractDownloadDestination`. The current interface is fairly limited though and I am not sure if it provides everything that you actually need.

For instance, I would imagine that you will want to decouple the network readahead buffer from the actual read size. File objects are optimized for returning data that maybe needs several calls to the underlying layer, while in your case an opposite approach may be necessary: you want the underlying layer to perform one big read, while the data should be handed to your layer in multiple chunks as they arrive. For example, you'd like to read 1MB into memory, but return first 64KB (which the user has requested) as soon as possible (in such case lets assume that the latency is high and bandwidth is low).

**Comment by ppolewicz** [comment]
ping @sondree

**Comment by sondree** [comment]
Yes, you are spot on in your assumptions. I do think that we can limit
B2Fuse to a bit higher minimum required bandwidth. At less than say
512kbytes/s, B2Fuse will not behave well.

I am unsure whether the reduction of file to parts download, should happen
in B2cli api or in B2Fuse. As it is now (in my beta) the smallest chunk
size I ever download is 1MB. I intend to keep this size fairly high as
significant additional bookkeeping is required per file chunk. The idea is
just to make B2Fuse more responsive.

As to answer the initial question, on what would be a good interface, that
is in large up to you. I need a way to either get feedback during a
download if B2Cli gets entire files at once or download chunks of files. My
current version uses partial downloads, but I can adapt to both.

If I did not mention this to you already I have a minimal working version
using B2 Cli Api for everything but file content updates.

Hope that answers your question, if not let me know.

Best
Sondre

**Comment by sondree** [comment]
One additinal small comment, please feel free to see how this fits with
your development path forward. Dual use of code rather than specialization
would be beneficial.

**Comment by ppolewicz** [comment]
Could you link to the implementation you mentioned so that we can see how you use our API? It could help us a lot to understand how you are working with it.

**Comment by ppolewicz** [comment]
I implemented most of https://github.com/Backblaze/B2_Command_Line_Tool/issues/250#issuecomment-249288739, still working on tests.

**Comment by sondree** [comment]
I am away from home and do not have access to my code right now. I'll make it available once I am back home (probably early next week). Essentially I am using the following functions:

> B2Api.get_bucket_by_id(...)
> Bucket.list_file_names(...)

I will be using these for basic file handling: 

> Bucket.delete_file_version(...)
> Bucket,download_file_by_id(...)
> Bucket.upload(...)

Basic support should be easy, what I have spent most of the time with B2Fuse on is getting acceptable performance and maintaining consistency.

Great! With this feature I believe we have everything required for me to complete a first release of B2Fuse based on your B2 Python API.

For each open file I will be creating a local "File" like object. This will be responsible for consistency and also background download (if required) of file content. This is essentially a wrapper around the implementation from this feature request. In addition this wrapper will handle the fact that files may be sparse and have holes where data is not yet available.

Best regards

**Comment by sondree** [comment]
So here it goes https://github.com/sondree/b2_fuse/tree/release1.3-beta

First version with a partial port to use B2 Python API.

Current implementation includes:

- Utilizing a bucket with B2 CLI API
- Listing of files with B2 CLI API
- Deletion of files using B2 CLI API

In the works:

- Upload data using B2 CLI API
- Download data using B2 CLI API


This version will be horribly slow as there is no caching structure yet built upon B2 CLI API. Essentially a bare minimum for this to be faster would be to cache "list_file_names" and invalidate cache on upload or delete. This would significantly increase the performance. If you want to try it out I suggest a custom bucket with less than 10 files.

I would like your opinion on the topic of caching. Could I use _B2.cache_ to cache the file listing as mentioned above? It would not be an issue to implement this on my side either. As I see it there are four options:

1. Using B2 Python API's caching structure
2. Wrapping/extending the B2 Python API's base bucket to implement a cache (which is what I do using B2Fuse base bucket today)
3. Separating caching and delegating this to B2Fuse itself (as B2Fuse already is the place concerned with consistency of the online file system).
4. B2Fuse providing a caching object to B2 CLI which is queried before a value is returned. 

Moving forward the plan is to make two B2Fuse "file types" B2SparseFileMemory and B2SparseFileDisk. These should in theory cover most use cases. For each open file in the FUSE driver I will create an object handling the download and upload of data providing a customized interface to B2Fuse. The B2Fuse "file types" will wrap the file objects as discussed in this issue. 

Best

**Comment by ppolewicz** [comment]
`b2.cache` is for something else. You should not use it for this purpose.

There are caching decorators in standard library (`lru_cache` since 3.2, other packages with wider compatibility) that you might use. I suggest that you inherit from bucket which will contain the caching and invalidation in the subclass. The basic version could just dump the cache after delete or upload and the more advanced version could also amend the cache if upload/delete was a success. Even more advanced version could "understand" the file "structure" that we have to serve a subdirectory listing from a recursive listing or to provide a negative confirmation (reply when someone is trying to list a directory that doesn't exist) etc.

**Comment by sondree** [comment]
Sounds good, this is the same way I was doing it previously. I have implemented a `CachedBucket` extending the default `b2.bucket`. Next up is an update to directory structure, as this is required to work with the B2 CLI API downloads files. 

What is the proper way to upload a using your API?

**Comment by ppolewicz** [comment]
```python
bucket.upload_local_file(path, 'remote_filename')
```
or
```python
bucket.upload_bytes(data, 'remote_filename')
```
There is upload progress reporting capability, but I am not sure if you are interested in that a this stage.
